src/basic/fs-util.c

```diff
@@ -1351,7 +1354,45 @@ int unlinkat_deallocate(int fd, const char *name, int flags) {
                 return 0;
         }
 
-        if (!S_ISREG(st.st_mode) || st.st_blocks == 0 || st.st_nlink > 0)
+        if (!S_ISREG(st.st_mode))
+                return 0;
+
+        if (FLAGS_SET(flags, UNLINK_ERASE) && st.st_size > 0 && st.st_nlink == 0) {
+                uint64_t left = st.st_size;
+                char buffer[64 * 1024];
+
+                /* If erasing is requested, let's overwrite the file with random data once before deleting
+                 * it. This isn't going to give you shred(1) semantics, but hopefully should be good enough
+                 * for stuff backed by tmpfs at least.
+                 *
+                 * Note that we only erase like this if the link count of the file is zero. If it is higer it
+                 * is still linked by someone else and we'll leave it to them to remove it securely
+                 * eventually! */
+
+                random_bytes(buffer, sizeof(buffer));
```

Comment:
we don't care about the quality much. that's why we use random_bytes() instead of genuine_random_bytes()... random_bytes() after all tries to get data from kernel if it can and doesn't mean blocking and uses genuine_random_bytes() otherwise.

But you are right we could no bother with the kernel random buffer at all here, and do entirely pseudo-random stuff here. But it appeared to me this would be unnecessarily sloppy. I mean, we do this to flush out crypto keys, and if it's no real difference to us it appears to still be better to overwrite with actual random stuff if we can get it cheaply than guessable pseudo-random stuff always.

I mean, in a way this is supposed to be as close to what shred(1) does is we can, except it should be graceful and quick and not be needlessly paranoid. shred() overwrites everything 3 times or so with random data generated from /dev/urandom. Now I am not sure we want to spend the 3 times thing, we want to be quick after all, but I'd still do the /dev/urandom thing if it's cheap and easy to do. And it is for us, it's just a slightly different function we have to call... Hope that makes some sense?